### PR TITLE
requestDevice refactor

### PR DESCRIPTION
**Pull request by zakorgy**
Refactor requestDevice function according to the specification changes. Two new members in `BluetoothRequestDeviceFilter` and one new member in `RequestDeviceOptions`.
Moved the `request_bluetooth_devices` algorithm out from the `requestDevice` function.
Also added annotations to the related functions.
Links to the spec: https://webbluetoothcg.github.io/web-bluetooth/#device-discovery, https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetooth-requestdevice, https://webbluetoothcg.github.io/web-bluetooth/#request-bluetooth-devices,
https://webbluetoothcg.github.io/web-bluetooth/#matches-a-filter

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #12614

- [x] These changes do not require tests because , there is no WebBluetooth Test API implementation yet.

### CODE REVIEW

File: components/net_traits/bluetooth_scanfilter.rs [hunk 1]
```diff
@@ -51,8 +60,20 @@ impl BluetoothScanfilter {
...
+        &self.service_data_uuid
+    }
+
     pub fn is_empty_or_invalid(&self) -> bool {
-        (self.name.is_empty() && self.name_prefix.is_empty() && self.get_services().is_empty()) ||
+        (self.name.is_empty() &&
+         self.name_prefix.is_empty() &&
+         self.get_services().is_empty() &&
+         self.manufacturer_id.is_none() &&
+         self.service_data_uuid.is_empty()) ||
```
Comment (mmatyas) [inline components/net_traits/bluetooth_scanfilter.rs:76]: this indentation looks interesting

File: components/script/dom/bluetooth.rs [hunk 2]
```diff
@@ -15,13 +15,13 @@ use dom::bindings::reflector::{Reflectable, Reflector, reflect_dom_object};
...
 use dom::bluetoothdevice::BluetoothDevice;
-use dom::bluetoothuuid::BluetoothUUID;
+use dom::bluetoothuuid::{BluetoothUUID, BluetoothServiceUUID};
 use ipc_channel::ipc::{self, IpcSender};
 use net_traits::bluetooth_scanfilter::{BluetoothScanfilter, BluetoothScanfilterSequence};
 use net_traits::bluetooth_scanfilter::{RequestDeviceoptions, ServiceUUIDSequence};
 use net_traits::bluetooth_thread::{BluetoothError, BluetoothMethodMsg};
 
-const FILTER_EMPTY_ERROR: &'static str = "'filters' member must be non - empty to find any devices.";
+const FILTER_EMPTY_ERROR: &'static str = "'filters' member, if present, must be non - empty to find any devices.";
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: `non - empty` is one word

File: components/script/dom/bluetooth.rs [hunk 3]
```diff
@@ -6,7 +6,7 @@ use bluetooth_blacklist::{Blacklist, uuid_is_blacklisted};
 use core::clone::Clone;
 use dom::bindings::codegen::Bindings::BluetoothBinding;
 use dom::bindings::codegen::Bindings::BluetoothBinding::RequestDeviceOptions;
-use dom::bindings::codegen::Bindings::BluetoothBinding::{BluetoothScanFilter, BluetoothMethods};
+use dom::bindings::codegen::Bindings::BluetoothBinding::{BluetoothRequestDeviceFilter, BluetoothMethods};
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: shouldn't this be `{BluetoothMethods, BluetoothRequestDeviceFilter}`

File: components/script/dom/bluetooth.rs [hunk 4]
```diff
@@ -34,6 +34,8 @@ const MAX_FILTER_NAME_LENGTH: usize = 29;
 const NAME_PREFIX_ERROR: &'static str = "'namePrefix', if present, must be non - empty.";
 const NAME_TOO_LONG_ERROR: &'static str = "A device name can't be longer than 248 bytes.";
 const SERVICE_ERROR: &'static str = "'services', if present, must contain at least one service.";
+const OPTIONS_ERROR: &'static str = "Fields of 'options' conflicts with each other.
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: `conflicts` -> `conflict`

File: components/net/bluetooth_thread.rs [hunk 5]
```diff
@@ -438,10 +461,19 @@ impl BluetoothManager {
...
-                                                           .collect();
+
+        // Step 6.
+        // Note: There is no requiredServiceUUIDS, we scan for all devices.
+        let mut matched_devices = self.get_and_cache_devices(&mut adapter);
+
+        // Step 7.
+        if !options.is_accepting_all_services() {
+            matched_devices = matched_devices.into_iter()
+                                     .filter(|d| matches_filters(d, options.get_filters()))
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: indentation

File: components/script/dom/bluetooth.rs [hunk 6]
```diff
@@ -150,24 +264,18 @@ impl From<BluetoothError> for Error {
...
+           (option.filters.is_none() && !option.acceptAllDevices) {
+            return Err(Type(OPTIONS_ERROR.to_owned()));
         }
+        // Step 2.
+        if !option.acceptAllDevices {
+            return self.request_bluetooth_devices(&option.filters, &option.optionalServices);
+        }
+
+        self.request_bluetooth_devices(&None, &option.optionalServices)
+        // TODO(#4282): Step 2-4: Reject and resolve promise.
```
Comment (zakorgy) [inline components/script/dom/bluetooth.rs]: Step 3-5

File: components/script/dom/bluetooth.rs [hunk 7]
```diff
@@ -59,41 +61,151 @@ impl Bluetooth {
...
+    fn request_bluetooth_devices(&self,
+                                 filters: &Option<Vec<BluetoothRequestDeviceFilter>>,
+                                 optional_services: &Option<Vec<BluetoothServiceUUID>>)
+                                 -> Fallible<Root<BluetoothDevice>> {
+        // TODO: Step 1: Triggered by user activation.
+
+        // Step 2.
+        let option = try!(convert_request_device_options(filters, optional_services, self.global().r()));
+
+        // TODO: Step 3-5: Impelemnt the permission API.
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: Impelemnt
whatshgfj ishfjgh thisjrej

File: components/script/dom/bluetooth.rs [hunk 8]
```diff
@@ -59,41 +61,151 @@ impl Bluetooth {
...
+
+        // Note: Steps 6-8 are implemented in
+        // components/net/bluetooth_thread.rs in request_device function.
+        let (sender, receiver) = ipc::channel().unwrap();
+        self.get_bluetooth_thread().send(BluetoothMethodMsg::RequestDevice(option, sender)).unwrap();
+        let device = receiver.recv().unwrap();
+
+        // TODO: Step 9-10: Impelemnt the permission API.
+
+        // Step 11: This step is optionable.
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: `optionable` -> `optional`

File: components/script/dom/bluetooth.rs [hunk 9]
```diff
@@ -59,41 +61,151 @@ impl Bluetooth {
...
+        }
+
+    }
 }
 
-fn canonicalize_filter(filter: &BluetoothScanFilter, global: GlobalRef) -> Fallible<BluetoothScanfilter> {
-    if filter.services.is_none() && filter.name.is_none() && filter.namePrefix.is_none() {
-        return Err(Type(FILTER_ERROR.to_owned()));
+// https://webbluetoothcg.github.io/web-bluetooth/#request-bluetooth-devices
+fn convert_request_device_options(filters: &Option<Vec<BluetoothRequestDeviceFilter>>,
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: shouldn't `global: GlobalRef` be the first param?

File: components/script/dom/bluetooth.rs [hunk 10]
```diff
@@ -59,41 +61,151 @@ impl Bluetooth {
...
+                                  optional_services: &Option<Vec<BluetoothServiceUUID>>,
+                                  global: GlobalRef)
+                                  -> Fallible<RequestDeviceoptions> {
+    // Step 2.2: There is no requiredServiceUUIDS, we scan for all devices.
+    let mut uuid_filters = vec!();
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs:112]: can we optimize this so `uuid_filters` is created only if `!filters.is_empty()`?

File: components/script/dom/bluetooth.rs [hunk 11]
```diff
@@ -59,41 +61,151 @@ impl Bluetooth {
...
+    }
+
+    let mut optional_services_uuids = vec!();
+    if let &Some(ref opt_services) = optional_services {
+        for opt_service in opt_services {
+            // Step 2.5 - 2.6.
+            let uuid = try!(BluetoothUUID::GetService(global, opt_service.clone())).to_string();
+
+            // Step 2.7.
+            // Note: What we doing here is adding the not blacklisted UUIDs to the result vector,
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: `doing` -> `are doing`

File: components/script/dom/bluetooth.rs [hunk 12]
```diff
@@ -103,36 +215,38 @@ fn canonicalize_filter(filter: &BluetoothScanFilter, global: GlobalRef) -> Falli
...
+    if let Some(manufact_id) = filter.manufacturerId {
+        manufacturer_id = Some(manufact_id);
     }
 
-    let mut filters = vec!();
-    for filter in &options.filters {
-        filters.push(try!(canonicalize_filter(&filter, global)));
-    }
+    let mut service_data_uuid = String::new();
+    // Step 2.4.7.
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs:231]: I'd prefer the `// Step` comments above the related variable, eg. above `service_data_uuid`.

File: components/script/dom/bluetooth.rs [hunk 13]
```diff
@@ -103,36 +215,38 @@ fn canonicalize_filter(filter: &BluetoothScanFilter, global: GlobalRef) -> Falli
...
+    if let Some(ref sdu) = filter.serviceDataUUID {
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: maybe `sduuid`?

File: tests/html/bluetooth/bluetooth_request_all_device.html [hunk 14]
```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE html>
```
Comment (mmatyas) [inline tests/html/bluetooth/bluetooth_request_all_device.html]: All Device -> Add Devices, in filename too

File: tests/html/bluetooth/bluetooth_request_all_device.html [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+        var options = {optionalServices: ['generic_acces'], acceptAllDevices:true, };
+        try {
+            log('Requesting Bluetooth Device...');
+            var bluetooth = window.navigator.bluetooth;
+            var device = bluetooth.requestDevice(options);
+
+            log('Connecting to GATT Server on device...');
+            var server = device.gatt.connect();
+
+            log('Getting Generic Acces Service...');
```
Comment (mmatyas) [inline tests/html/bluetooth/bluetooth_request_all_device.html]: `Acces` typo

File: tests/html/bluetooth/bluetooth_request_all_device.html [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+<html>
+<title>Request All Device</title>
+<body>
+    <button type="button" onclick="onButtonClick()">Request All Device</button>
+    <pre id="log"></pre>
+    <script src="bluetooth_functions.js"></script>
+    <script>
+    function onButtonClick() {
+        clear();
+        var options = {optionalServices: ['generic_acces'], acceptAllDevices:true, };
```
Comment (mmatyas) [inline tests/html/bluetooth/bluetooth_request_all_device.html]: `acces` typo

File: components/net/bluetooth_thread.rs [hunk 17]
```diff
@@ -115,6 +119,24 @@ fn matches_filter(device: &BluetoothDevice, filter: &BluetoothScanfilter) -> boo
...
+// Step 4.
+// TODO: Implement get_manufacturer_data to device.
+//    if let Some(manufacturer_id) = filter.get_manufacturer_id() {
+//        if !device.get_manufacturer_data().contains_key(manufacturer_id) {
+//            return false;
+//        }
+//    }
+//
+// Step 5.
+// TODO: Implement get_device_data to device.
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: `Implement get_device_data to device.` -> `Implement get_device_data in device crate.`

File: components/net/bluetooth_thread.rs [hunk 18]
```diff
@@ -115,6 +119,24 @@ fn matches_filter(device: &BluetoothDevice, filter: &BluetoothScanfilter) -> boo
             }
         }
     }
+
+// Step 4.
+// TODO: Implement get_manufacturer_data to device.
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: `Implement get_manufacturer_data to device.` -> `Implement get_manufacturer_data in device crate.`

File: components/net_traits/bluetooth_scanfilter.rs (hunk 1 above)
Comment (dati91) [reply components/net_traits/bluetooth_scanfilter.rs:76]: it looks just fine

File: components/script/dom/bluetooth.rs [hunk 19]
```diff
@@ -15,13 +15,13 @@ use dom::bindings::reflector::{Reflectable, Reflector, reflect_dom_object};
 use dom::bindings::str::DOMString;
 use dom::bluetoothadvertisingdata::BluetoothAdvertisingData;
 use dom::bluetoothdevice::BluetoothDevice;
-use dom::bluetoothuuid::BluetoothUUID;
+use dom::bluetoothuuid::{BluetoothUUID, BluetoothServiceUUID};
```
Comment (dati91) [inline components/script/dom/bluetooth.rs]: maybe `{BluetoothServiceUUID, BluetoothUUID}`, but I'm not sure about this.

File: components/script/dom/bluetooth.rs (hunk 10 above)
Comment (dati91) [reply components/script/dom/bluetooth.rs:112]: we will need an empty vec, so no

File: components/script/dom/bluetooth.rs [hunk 20]
```diff
@@ -59,41 +61,151 @@ impl Bluetooth {
...
+       filter.name.is_none() &&
+       filter.namePrefix.is_none() &&
+       filter.manufacturerId.is_none() &&
+       filter.serviceDataUUID.is_none() {
+           return Err(Type(FILTER_ERROR.to_owned()));
+    }
+
+    // Step 2.4.2: There is no empty canonicalizedFilter member,
+    // we create a BluetoothScanfilter instance at the end of the function.
+
```
Comment (dati91) [inline components/script/dom/bluetooth.rs]: remove empty line

File: components/script/dom/bluetooth.rs (hunk 12 above)
Comment (dati91) [reply components/script/dom/bluetooth.rs:231]: You can do `let var = match function {...}` to initialize a variable.

File: tests/html/bluetooth/bluetooth_request_all_device.html (hunk 14 above)
Comment (zakorgy) [reply tests/html/bluetooth/bluetooth_request_all_device.html]: You mean All Devices?

Comment (mmatyas) [reply tests/html/bluetooth/bluetooth_request_all_device.html]: Yup

File: components/script/dom/webidls/Bluetooth.webidl [hunk 21]
```diff
@@ -4,24 +4,31 @@
...
+//  [SecureContext]
+//  Promise<BluetoothDevice> requestDevice(RequestDeviceOptions options);
+  [Throws]
+  BluetoothDevice requestDevice(optional RequestDeviceOptions options);
 };
 
-// Bluetooth implements EventTarget;
-// Bluetooth implements CharacteristicEventHandlers;
-// Bluetooth implements ServiceEventHandlers;
+//Bluetooth implements EventTarget;
```
Comment (dati91) [inline components/script/dom/webidls/Bluetooth.webidl]: the previous commenting was better

File: tests/html/bluetooth/bluetooth_request_all_devices.html [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+<html>
+<title>Request All Devices</title>
+<body>
+    <button type="button" onclick="onButtonClick()">Request All Devices</button>
+    <pre id="log"></pre>
+    <script src="bluetooth_functions.js"></script>
+    <script>
+    function onButtonClick() {
+        clear();
+        var options = {optionalServices: ['generic_access'], acceptAllDevices:true, };
```
Comment (dati91) [inline tests/html/bluetooth/bluetooth_request_all_devices.html]: remove the last `,`

**Pull request merged**